File: src/services/screenshotService.js
const puppeteer = require('puppeteer');
const { withTimeout, forceBrowserClose, getTimeoutFromEnv } = require('../utils/timeout');

class ScreenshotService {
  getBrowserOptions() {
    const options = {
      headless: "new",
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage'
      ]
    };

    // Use executable path from env if provided (for Docker)
    if (process.env.PUPPETEER_EXECUTABLE_PATH) {
      options.executablePath = process.env.PUPPETEER_EXECUTABLE_PATH;
    }

    return options;
  }

  async captureScreenshot(url, options = {}) {
    // Extract options with defaults
    const {
      fullPage = true,
      format = 'png',
      quality = 90,
      delay = 0,
      viewport = { width: 1920, height: 1080 }
    } = options;

    // Validate format
    if (!['png', 'jpeg', 'webp'].includes(format)) {
      throw new Error('Invalid format. Must be png, jpeg, or webp');
    }

    // Validate quality (only for jpeg and webp)
    if ((format === 'jpeg' || format === 'webp') && (quality < 1 || quality > 100)) {
      throw new Error('Quality must be between 1 and 100');
    }

    // Validate delay
    if (delay < 0 || delay > 30000) {
      throw new Error('Delay must be between 0 and 30000 milliseconds');
    }

    // Get configurable timeouts from environment
    const OVERALL_TIMEOUT = getTimeoutFromEnv('SCREENSHOT_OPERATION_TIMEOUT', 80000);
    const BROWSER_LAUNCH_TIMEOUT = getTimeoutFromEnv('SCREENSHOT_BROWSER_LAUNCH_TIMEOUT', 15000);
    const PAGE_NAVIGATION_TIMEOUT = getTimeoutFromEnv('SCREENSHOT_PAGE_NAVIGATION_TIMEOUT', 30000);
    const SCREENSHOT_CAPTURE_TIMEOUT = getTimeoutFromEnv('SCREENSHOT_CAPTURE_TIMEOUT', 20000);
    const SCREENSHOT_GCS_UPLOAD_TIMEOUT = getTimeoutFromEnv('SCREENSHOT_GCS_UPLOAD_TIMEOUT', 15000);

    console.log('Configured timeouts (ms):', {
      OVERALL: OVERALL_TIMEOUT,
      BROWSER_LAUNCH: BROWSER_LAUNCH_TIMEOUT,
      PAGE_NAVIGATION: PAGE_NAVIGATION_TIMEOUT,
      SCREENSHOT_CAPTURE: SCREENSHOT_CAPTURE_TIMEOUT,
      GCS_UPLOAD: SCREENSHOT_GCS_UPLOAD_TIMEOUT
    });

    // Store browser reference for cleanup on timeout
    let browser = null;

    // Wrap entire operation with overall timeout
    return withTimeout(
      this._captureScreenshotInternal(
        url,
        { fullPage, format, quality, delay, viewport },
        BROWSER_LAUNCH_TIMEOUT,
        PAGE_NAVIGATION_TIMEOUT,
        SCREENSHOT_CAPTURE_TIMEOUT,
        (b) => { browser = b; }
      ),
      OVERALL_TIMEOUT,
      'Screenshot capture operation',
      async () => {
        // Cleanup function called on overall timeout
        console.error('Overall screenshot operation timed out, force-closing browser');
        if (browser) {
          await forceBrowserClose(browser);
        }
      }
    );
  }

  async _captureScreenshotInternal(
    url,
    options,
    browserLaunchTimeout,
    pageNavigationTimeout,
    screenshotCaptureTimeout,
    setBrowser
  ) {
    const { fullPage, format, quality, delay, viewport } = options;

    // Get wait strategy from environment (default: load - waits for all resources)
    const waitStrategy = process.env.SCREENSHOT_WAIT_STRATEGY || 'load';

    // Get configurable post-load delay (default: 2000ms for JS execution)
    const postLoadDelay = parseInt(process.env.SCREENSHOT_POST_LOAD_DELAY || '2000', 10);

    console.log(`Launching browser with ${browserLaunchTimeout}ms timeout...`);

    // Launch browser with timeout
    const browser = await withTimeout(
      puppeteer.launch(this.getBrowserOptions()),
      browserLaunchTimeout,
      'Browser launch'
    );

    // Store browser reference for cleanup
    setBrowser(browser);

    console.log('Browser launched successfully');

    try {
      const memoryBefore = process.memoryUsage();
      console.log(`Memory before page creation: RSS=${Math.round(memoryBefore.rss / 1024 / 1024)}MB, Heap=${Math.round(memoryBefore.heapUsed / 1024 / 1024)}MB`);

      const page = await browser.newPage();

      // Set viewport
      await page.setViewport(viewport);

      console.log(`Navigating to ${url} with waitUntil: ${waitStrategy} (timeout: ${pageNavigationTimeout}ms)...`);

      // Navigate to URL with appropriate wait strategy
      await withTimeout(
        page.goto(url, {
          waitUntil: waitStrategy,
          timeout: pageNavigationTimeout
        }),
        pageNavigationTimeout,
        'Page navigation'
      );

      console.log('Page loaded successfully');

      // Scroll the page to trigger lazy-loaded content
      console.log('Scrolling page to trigger lazy-loaded content...');
      await this._autoScroll(page).catch(err => {
        console.warn('Auto-scroll completed with warnings:', err.message);
      });

      // Wait for images to load (after scrolling)
      console.log('Waiting for images to load...');
      await this._waitForImages(page).catch(err => {
        console.warn('Image loading wait completed with warnings:', err.message);
      });

      // Post-load delay for JavaScript execution and dynamic content
      const totalDelay = Math.max(delay || 0, postLoadDelay);
      if (totalDelay > 0) {
        const safeDelay = Math.min(totalDelay, 10000);
        console.log(`Waiting ${safeDelay}ms for dynamic content and JS execution...`);
        await new Promise(resolve => setTimeout(resolve, safeDelay));
      }

      // Capture screenshot with timeout
      const screenshotOptions = {
        fullPage,
        type: format
      };

      // Add quality for jpeg and webp
      if (format === 'jpeg' || format === 'webp') {
        screenshotOptions.quality = quality;
      }

      console.log(`Capturing screenshot (timeout: ${screenshotCaptureTimeout}ms)...`);

      const buffer = await withTimeout(
        page.screenshot(screenshotOptions),
        screenshotCaptureTimeout,
        'Screenshot capture'
      );

      console.log(`Screenshot captured successfully (${buffer.length} bytes)`);

      const memoryAfter = process.memoryUsage();
      console.log(`Memory after screenshot: RSS=${Math.round(memoryAfter.rss / 1024 / 1024)}MB, Heap=${Math.round(memoryAfter.heapUsed / 1024 / 1024)}MB`);

      // Graceful close with force-kill fallback
      await forceBrowserClose(browser, 3000);

      return buffer;
    } catch (error) {
      console.error('Screenshot capture error:', error.message);

      // Aggressive cleanup on error
      await forceBrowserClose(browser, 3000);

      throw error;
    }
  }

  /**
   * Auto-scroll the page to trigger lazy-loaded content
   * @param {Object} page - Puppeteer page object
   */
  async _autoScroll(page) {
    console.log('Starting page auto-scroll...');
    try {
      await page.evaluate(async () => {
        await new Promise((resolve) => {
          let totalHeight = 0;
          const distance = 100;
          const delay = 50; // Faster scroll
          const maxHeight = 15000; // Cap at 15k pixels to prevent infinite loops
          const scrollTimeout = 40000; // 40s max for scrolling
          const startTime = Date.now();

          const timer = setInterval(() => {
            const scrollHeight = document.body.scrollHeight;
            window.scrollBy(0, distance);
            totalHeight += distance;

            const timeElapsed = Date.now() - startTime;

            // Stop when we've scrolled to the bottom, reached max height, or timed out
            if (totalHeight >= scrollHeight - window.innerHeight || totalHeight >= maxHeight || timeElapsed >= scrollTimeout) {
              clearInterval(timer);
              window.scrollTo(0, 0);
              resolve();
            }
          }, delay);
        });
      });

      console.log('Page auto-scroll completed');
    } catch (error) {
      console.warn('Error during auto-scroll:', error.message);
      throw error;
    }
  }

  /**
   * Wait for images to load on the page
   * @param {Object} page - Puppeteer page object
   * @param {number} timeout - Timeout in milliseconds
   */
  async _waitForImages(page, timeout = 5000) {
    try {
      await page.evaluate(async () => {
        const images = Array.from(document.querySelectorAll('img'));

        await Promise.all(
          images.map(img => {
            if (img.complete) return Promise.resolve();

            return new Promise((resolve, reject) => {
              img.addEventListener('load', resolve);
              img.addEventListener('error', resolve); // Resolve even on error to not block

              // Timeout for individual image
              setTimeout(resolve, 3000);
            });
          })
        );
      });

      console.log('All images loaded successfully');
    } catch (error) {
      console.warn('Error waiting for images:', error.message);
      throw error;
    }
  }
}

module.exports = new ScreenshotService();

File: app/services/screenshot_service.py
import asyncio
import logging
import time

from app.config import settings
from app.services.browser_pool import browser_pool

logger = logging.getLogger(__name__)

# SPA-aware content readiness check
WAIT_FOR_CONTENT_JS = """async () => {
    const maxWait = 20000;
    const start = Date.now();

    const SPA_SELECTORS = ['#root', '#app', '#__next', '#__nuxt', '[data-reactroot]', 'main'];

    while (Date.now() - start < maxWait) {
        // Check SPA root containers for rendered children with real content
        for (const sel of SPA_SELECTORS) {
            const el = document.querySelector(sel);
            if (el && el.children.length > 0) {
                const text = el.innerText ? el.innerText.trim() : '';
                const imgs = el.querySelectorAll('img[src]:not([src=""])');
                if (text.length > 50 || imgs.length > 1) {
                    // Found a rendered SPA root — wait 1s more for async data
                    await new Promise(r => setTimeout(r, 1000));
                    const finalText = el.innerText ? el.innerText.trim() : '';
                    const finalImgs = el.querySelectorAll('img[src]:not([src=""])');
                    return {ready: true, textLen: finalText.length, imgCount: finalImgs.length, source: sel};
                }
            }
        }

        // Fallback: check body for meaningful content (higher threshold)
        const bodyText = document.body ? document.body.innerText.trim() : '';
        const bodyImgs = document.querySelectorAll('img[src]:not([src=""])');
        if (bodyText.length > 200 || bodyImgs.length > 3) {
            return {ready: true, textLen: bodyText.length, imgCount: bodyImgs.length, source: 'body'};
        }

        await new Promise(r => setTimeout(r, 500));
    }
    // Timed out — return whatever state we have
    const text = document.body ? document.body.innerText.trim() : '';
    return {ready: false, textLen: text.length, imgCount: document.querySelectorAll('img').length, source: 'timeout'};
}"""

AUTO_SCROLL_JS = """async () => {
    await new Promise((resolve) => {
        const viewportHeight = window.innerHeight;
        const distance = Math.floor(viewportHeight * 0.8);
        const maxHeight = 15000;
        const scrollTimeout = 40000;
        const startTime = Date.now();
        let totalHeight = 0;

        const timer = setInterval(() => {
            const scrollHeight = document.body.scrollHeight;
            window.scrollBy(0, distance);
            totalHeight += distance;

            if (totalHeight >= scrollHeight - viewportHeight || totalHeight >= maxHeight || Date.now() - startTime >= scrollTimeout) {
                clearInterval(timer);
                window.scrollTo(0, 0);
                resolve();
            }
        }, 200);
    });
}"""

WAIT_FOR_IMAGES_JS = """async () => {
    const images = Array.from(document.querySelectorAll('img'));
    await Promise.all(images.map(img => {
        if (img.complete) return;
        return new Promise(resolve => {
            img.addEventListener('load', resolve);
            img.addEventListener('error', resolve);
            setTimeout(resolve, 3000);
        });
    }));
}"""


async def capture_screenshot(
    url: str,
    full_page: bool = True,
    fmt: str = "png",
    quality: int = 90,
    delay: int = 0,
) -> bytes:
    overall_timeout = settings.screenshot_operation_timeout / 1000

    async def _do_capture() -> bytes:
        context = await browser_pool.acquire_context()
        try:
            page = await context.new_page()

            nav_timeout = settings.screenshot_page_navigation_timeout
            logger.info("Navigating to %s (timeout: %dms)...", url, nav_timeout)

            # Use commit — fires earliest, we handle waiting ourselves
            await page.goto(url, wait_until="commit", timeout=nav_timeout)
            logger.info("Navigation committed, waiting for load state...")

            # Wait for load state so JS bundles are fetched
            try:
                await page.wait_for_load_state("load", timeout=30000)
                logger.info("Load state reached")
            except Exception:
                logger.info("Load state timeout (30s), continuing...")

            # Wait for real visible content (SPA-aware)
            logger.info("Waiting for visible content to render...")
            content_state = await page.evaluate(WAIT_FOR_CONTENT_JS)
            logger.info("Content state: ready=%s, text=%d chars, images=%d, source=%s",
                        content_state.get("ready"), content_state.get("textLen", 0),
                        content_state.get("imgCount", 0), content_state.get("source", ""))

            # If content not ready, try networkidle then re-check
            if not content_state.get("ready"):
                logger.info("Content not ready, trying networkidle...")
                try:
                    await page.wait_for_load_state("networkidle", timeout=10000)
                except Exception:
                    pass
                content_state = await page.evaluate(WAIT_FOR_CONTENT_JS)
                logger.info("Content state after networkidle: ready=%s, text=%d chars, images=%d",
                            content_state.get("ready"), content_state.get("textLen", 0),
                            content_state.get("imgCount", 0))

            # Auto-scroll to trigger lazy loaders
            logger.info("Scrolling page...")
            await page.evaluate(AUTO_SCROLL_JS)

            # Wait for network to settle after scrolling
            try:
                await page.wait_for_load_state("networkidle", timeout=10000)
            except Exception:
                pass

            # Wait for images to finish loading
            await page.evaluate(WAIT_FOR_IMAGES_JS)

            # Post-load delay
            post_load = settings.screenshot_post_load_delay
            total_delay = max(delay, post_load)
            if total_delay > 0:
                safe_delay = min(total_delay, 10000)
                logger.info("Waiting %dms for dynamic content...", safe_delay)
                await asyncio.sleep(safe_delay / 1000)

            # Capture
            capture_timeout_ms = settings.screenshot_capture_timeout
            screenshot_opts: dict = {
                "full_page": full_page,
                "type": fmt,
                "timeout": capture_timeout_ms,
            }
            if fmt in ("jpeg", "webp"):
                screenshot_opts["quality"] = quality

            logger.info("Capturing screenshot (timeout: %dms)...", capture_timeout_ms)
            buffer = await asyncio.wait_for(
                page.screenshot(**screenshot_opts),
                timeout=capture_timeout_ms / 1000,
            )

            logger.info("Screenshot captured (%d bytes)", len(buffer))
            return buffer
        finally:
            await browser_pool.release_context(context)

    start = time.time()
    try:
        result = await asyncio.wait_for(_do_capture(), timeout=overall_timeout)
        elapsed = int((time.time() - start) * 1000)
        logger.info("Total screenshot time: %dms", elapsed)
        return result
    except asyncio.TimeoutError:
        logger.error("Screenshot operation timed out after %ds", overall_timeout)
        raise
